### PR TITLE
Regularizer

### PR DESCRIPTION
**Pull request by MatheMax**
This works faster, what is not surprising but pretty nice!

I did not remove "SmoothnessN2()" until now. 
Probably we should do this to keep it simple.
@kkmann  What do you think?

### CODE REVIEW

File: R/Scores.R [hunk 1]
```diff
@@ -273,3 +273,76 @@ setMethod(".evaluate", signature("IntegralScore", "OneStageDesign"),
...
+                  return(res)
+              }
+          })
+
+
+# not user facing!
+setMethod(".evaluate", signature("SmoothnessL1", "TwoStageDesign"),
+          function(s, design, ...) {
+              sum(design@n2_pivots * design@weights) / (design@c1e - design@c1f)
+              }
```
Comment (kkmann) [inline R/Scores.R]: intendation

File: R/Scores.R [hunk 2]
```diff
@@ -273,3 +273,76 @@ setMethod(".evaluate", signature("IntegralScore", "OneStageDesign"),
...
+                  res <- res / (design@c1e - design@c1f)
+                  return(res)
+              }
+          })
+
+
+# not user facing!
+setMethod(".evaluate", signature("SmoothnessL1", "TwoStageDesign"),
+          function(s, design, ...) {
+              sum(design@n2_pivots * design@weights) / (design@c1e - design@c1f)
```
Comment (kkmann) [inline R/Scores.R]: shouldn't this use our integrate function to be generic?

File: tests/testthat/test_SmoothnessL1.R [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        evaluate(smth, design2),
+        2 * evaluate(smth, design1)
+    )
+
+    expect_equal( # everything is positive!
+        evaluate(smth, design3),
+        evaluate(smth, design1) + evaluate(smth, design2)
+    )
+}) # end 'SmoothnessL1 defines a norm'
+
```
Comment (kkmann) [inline tests/testthat/test_SmoothnessL1.R]: i think we have 3 blanks between test_that - makes it easier to switch cases quickly

File: R/Scores.R (hunk 2 above)
Comment (MatheMax) [reply R/Scores.R]: But this would become slowlier because n2(design, x) would be evaluated in any step

**Comment by MatheMax** [comment]
> 
> @kkmann wrote:
> Nice one, plz review code styling. More importantly we need a consistent class system. I do not see the need for a new abstract class between AverageN2 (more adequate name) and score. I am open to suggestions though if you have another good idea.

I would like to implement SmoothnessN2 also as a regularization subscore
Would be more consistent that the current implementation as IntegralScore

File: R/Scores.R (hunk 2 above)
Comment (kkmann) [reply R/Scores.R]: How much slower is that? My point is we should use a consistent integration rule - either always Gaussian quadrature or always simple average (as Stephan suggested!)

**Comment by kkmann** [comment]
> > @kkmann wrote:
> > Nice one, plz review code styling. More importantly we need a consistent class system. I do not see the need for a new abstract class between AverageN2 (more adequate name) and score. I am open to suggestions though if you have another good idea.
> 
> I would like to implement SmoothnessN2 also as a regularization subscore
> Would be more consistent that the current implementation as IntegralScore

But what's the difference between a score and a Regularizer? what additional structure does a Regularizer expose to warrant a new class?

File: R/Scores.R (hunk 2 above)
Comment (MatheMax) [reply R/Scores.R]: But this is Gaussian quadrature as the correct weights are used.

**Comment by kkmann** [comment]
I agree thought that SmoothnessN2 could also be implemented as direct subclass of Score (like AverageN2)

**Comment by MatheMax** [comment]
> 
> 
> I agree thought that SmoothnessN2 could also be implemented as direct subclass of Score (like AverageN2)

That's a deal

**Comment by kkmann** [comment]
if we are at it, it might also be worthwhile to have a N1 < Score (super trivial) that would allow to increase/decrease the weight on stage one sample size (not sure why, but hey - why not?)

I.e. ESS + .1 N1 would penalize large n1 and ESS - .1 N1 would emphasize larger n1

File: R/regularization.R [hunk 4]
```diff
@@ -1,3 +1,59 @@
...
+                  return(res)
+              }
+          }
+)
+
+
+# not user facing!
+setMethod(".evaluate", signature("AverageN2", "TwoStageDesign"),
+          function(s, design, ...) {
+              sum(design@n2_pivots * design@weights) / (design@c1e - design@c1f)
```
Comment (kkmann) [inline R/regularization.R]: still not happy with that (cf. above)

File: R/regularization.R [hunk 5]
```diff
@@ -68,7 +114,8 @@ setMethod("evaluate", signature("SmoothnessN2", "TwoStageDesign"),
                   return(1 / diff(x1_bounds) *
                              stats::integrate(integrand, x1_bounds[1], x1_bounds[2], subdivisions = 1000)$value)
```
Comment (kkmann) [inline R/regularization.R]: why do we need a maximum number of subdivisions here?

File: R/regularization.R [hunk 6]
```diff
@@ -1,3 +1,59 @@
+#' Regularization via L1 norm
```
Comment (kkmann) [inline R/regularization.R]: Maybe more specific docs? This IS going to come back at us at some point ;)

File: R/Scores.R (hunk 2 above)
Comment (kkmann) [reply R/Scores.R]: true but why exactly do you object to using the appropriate function then? It's good practice not to reimplement stuff twice - if u fix it in one place it's still broken in the other. Is this REALLY slower?

**Comment by kkmann** [comment]
excellent testing btw (100% coverage on diff!)

File: R/regularization.R (hunk 4 above)
Comment (MatheMax) [reply R/regularization.R]: I changed it, but if we get runtime problems, we can thinking of returning to this one

File: R/regularization.R (hunk 6 above)
Comment (MatheMax) [reply R/regularization.R]: https://github.com/kkmann/otsd/blob/05a05ebe3e04965acf67afae79621af6c2b4e917/R/regularization.R#L3-L9

File: R/regularization.R (hunk 5 above)
Comment (MatheMax) [reply R/regularization.R]: I think we don't and removed it.

https://github.com/kkmann/otsd/blob/05a05ebe3e04965acf67afae79621af6c2b4e917/R/regularization.R#L129

**Pull request merged**